subsys/bluetooth/host/buf.c

```diff
@@ -82,6 +82,96 @@ struct net_buf *bt_buf_get_rx(enum bt_buf_type type, k_timeout_t timeout)
 	return buf;
 }
 
+static bool is_hci_event_discardable(const uint8_t evt_type, const uint8_t subevt_type)
+{
+	switch (evt_type) {
+#if defined(CONFIG_BT_CLASSIC)
+	case BT_HCI_EVT_INQUIRY_RESULT_WITH_RSSI:
+	case BT_HCI_EVT_EXTENDED_INQUIRY_RESULT:
+		return true;
+#endif
+	case BT_HCI_EVT_LE_META_EVENT: {
+		switch (subevt_type) {
+		case BT_HCI_EVT_LE_ADVERTISING_REPORT:
+			return true;
+		default:
+			return false;
+		}
+	}
+	default:
+		return false;
+	}
+}
+
+NET_BUF_POOL_FIXED_DEFINE(sync_adv_pool, 1, SYNC_EVT_SIZE, sizeof(struct bt_buf_data), NULL);
+
+/* This is a host-private function */
+bool bt_buf_can_steal_ext_adv_report_buf(void)
+{
+	/* Host can steal the buffer if we have at least one available.
+	 * This is the case when:
+	 * - There are buffers in the `free` LIFO
+	 * - There are un-initialized buffers
+	 */
+	return (!k_fifo_is_empty(&sync_adv_pool.free)) || (sync_adv_pool.uninit_count != 0);
```

Comment:
This almost begs for a proper net_buf_pool API to test for "is it empty?". That said, it'd inherently racy in premptible contexts, so we might not want to give people such a foot gun.

Reply:
yep. Ideally `net_buf` would init all buffers on boot and place them in the `free` LIFO.